src/net_processing.cpp

```diff
@@ -186,6 +186,38 @@ struct QueuedBlock {
     std::unique_ptr<PartiallyDownloadedBlock> partialBlock;
 };
 
+template <size_t N>
+class TimeOffsets
+{
+private:
+    static_assert(N > 0, "TimeOffsets has to sample at least one element");
+
+    mutable Mutex m_mutex;
+    std::array<int64_t, N> m_offsets GUARDED_BY(m_mutex){};
+    size_t m_index GUARDED_BY(m_mutex){0};
+
+public:
+    void Add(int64_t offset) EXCLUSIVE_LOCKS_REQUIRED(!m_mutex)
+    {
+        AssertLockNotHeld(m_mutex);
+        LOCK(m_mutex);
+        m_offsets[m_index] = offset;
+        m_index = (1 + m_index) % N;
+    }
+    int64_t Median() const EXCLUSIVE_LOCKS_REQUIRED(!m_mutex)
+    {
+        AssertLockNotHeld(m_mutex);
+        std::array<int64_t, N> sorted_copy = WITH_LOCK(m_mutex, return m_offsets);
+        std::sort(sorted_copy.begin(), sorted_copy.end());
+
+        if constexpr (N % 2 == 0) {
+            return (sorted_copy[N / 2 - 1] + sorted_copy[N / 2]) / 2;
```

Comment:
122658f95b36b7f940351bd34a89e13831931f18

should we make this addition overflow safe?

Reply:
i won't mind that either, but ideally with an in-function comment.